pkg/cmd/server/kubernetes/node.go

```diff
@@ -192,77 +191,77 @@ func (c *NodeConfig) RunSDN() {
+	switch c.ProxyConfig.ProxyMode {
+	case "iptables":
+		glog.V(2).Info("Using iptables Proxier.")
+		proxierIptables, err := iptables.NewProxier(iptInterface, exec, c.ProxyConfig.IptablesSyncPeriod, c.ProxyConfig.MasqueradeAll)
+		if err != nil {
+			// This should be fatal, but that would break the integration tests
+			glog.Warningf("WARNING: Could not initialize Kubernetes Proxy. You must run this process as root to use the service proxy: %v", err)
+			return
+		}
+		proxier = proxierIptables
+		endpointsHandler = proxierIptables
+		// No turning back. Remove artifacts that might still exist from the userspace Proxier.
+		glog.V(2).Info("Tearing down userspace rules. Errors here are acceptable.")
+		userspace.CleanupLeftovers(iptInterface)
+	case "userspace":
+		glog.V(2).Info("Using userspace Proxier.")
+		loadBalancer := userspace.NewLoadBalancerRR()
 		endpointsHandler = loadBalancer
+		proxierUserspace, err := userspace.NewProxier(loadBalancer, c.ProxyConfig.BindAddress, iptInterface, c.ProxyConfig.PortRange, c.ProxyConfig.IptablesSyncPeriod, c.ProxyConfig.UDPIdleTimeout)
+		if err != nil {
+			// This should be fatal, but that would break the integration tests
+			glog.Warningf("WARNING: Could not initialize Kubernetes Proxy. You must run this process as root to use the service proxy: %v", err)
+			return
+		}
+		proxier = proxierUserspace
+		// Remove artifacts from the pure-iptables Proxier.
+		glog.V(2).Info("Tearing down pure-iptables proxy rules.")
+		iptables.CleanupLeftovers(iptInterface)
+	default:
+		glog.Fatalf("Unknown proxy mode %q", c.ProxyConfig.ProxyMode)
 	}
-	if err != nil {
-		// This should be fatal, but that would break the integration tests
-		glog.Warningf("WARNING: Could not initialize Kubernetes Proxy. You must run this process as root to use the service proxy: %v", err)
-		return
-	}
-	iptables.AddReloadFunc(proxier.Sync)
-
-	pconfig.NewSourceAPI(
-		c.Client,
-		10*time.Minute,
-		serviceConfig.Channel("api"),
-		endpointsConfig.Channel("api"))
+	iptInterface.AddReloadFunc(proxier.Sync)
 
+	// Create configs (i.e. Watches for Services and Endpoints)
+	// Note: RegisterHandler() calls need to happen before creation of Sources because sources
+	// only notify on changes, and the initial update (on process start) may be lost if no handlers
+	// are registered yet.
+	serviceConfig := pconfig.NewServiceConfig()
 	serviceConfig.RegisterHandler(proxier)
+	endpointsConfig := pconfig.NewEndpointsConfig()
 	if c.FilteringEndpointsHandler == nil {
 		endpointsConfig.RegisterHandler(endpointsHandler)
 	} else {
 		c.FilteringEndpointsHandler.SetBaseEndpointsHandler(endpointsHandler)
 		endpointsConfig.RegisterHandler(c.FilteringEndpointsHandler)
 	}
-	recorder.Eventf(nodeRef, kapi.EventTypeNormal, "Starting", "Starting kube-proxy.")
-	glog.Infof("Started Kubernetes Proxy on %s", host)
+
+	pconfig.NewSourceAPI(
+		c.Client,
+		c.ProxyConfig.ConfigSyncPeriod,
+		serviceConfig.Channel("api"),
+		endpointsConfig.Channel("api"))
```

Comment:
upstream code says you have to register handlers prior to adding the source so you don't miss anything... was there a reason we were doing that in the reverse order?